src/Symfony/Component/Serializer/DependencyInjection/RuntimeSerializerServicesPass.php

```diff
@@ -0,0 +1,157 @@
+
+use Symfony\Component\DependencyInjection\Attribute\Autowire;
+use Symfony\Component\DependencyInjection\Attribute\Target;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\Compiler\ServiceLocatorTagPass;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\DependencyInjection\TypedReference;
+use Symfony\Component\Serializer\Attribute\DeserializeFormatter;
+use Symfony\Component\Serializer\Attribute\MaxDepth;
+use Symfony\Component\Serializer\Attribute\SerializeFormatter;
+use Symfony\Component\Serializer\Deserialize\Config\DeserializeConfig;
+use Symfony\Component\Serializer\Serialize\Config\SerializeConfig;
+use Symfony\Component\VarExporter\ProxyHelper;
+
+/**
+ * Creates and injects a service locator containing serializable classes formatter's needed services.
+ *
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ *
+ * @experimental in 7.0
+ */
+final class RuntimeSerializerServicesPass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container): void
+    {
+        if (!$container->hasDefinition('serializer.serializer')) {
+            return;
+        }
+
+        $formatters = [];
+        foreach ($container->getDefinitions() as $definition) {
+            if (!$definition->hasTag('serializer.serializable')) {
+                continue;
+            }
+
+            array_push($formatters, ...$this->formatters($definition->getClass()));
+        }
+
+        $runtimeServices = [];
+        foreach ($formatters as $formatter) {
+            $formatterName = sprintf('%s::%s', $formatter->getClosureScopeClass()->getName(), $formatter->getName());
+            foreach ($this->retrieveServices($container, $formatter) as $serviceName => $reference) {
+                $runtimeServices[sprintf('%s[%s]', $formatterName, $serviceName)] = $reference;
+            }
+        }
+
+        $runtimeServicesLocator = ServiceLocatorTagPass::register($container, $runtimeServices);
+
+        $container->getDefinition('serializer.serializer')
+            ->replaceArgument(1, $runtimeServicesLocator);
+
+        $container->getDefinition('serializer.deserializer')
+            ->replaceArgument(1, $runtimeServicesLocator);
+
+        $container->getDefinition('serializer.serialize.data_model_builder')
+            ->replaceArgument(1, $runtimeServicesLocator);
+
+        $container->getDefinition('serializer.deserialize.data_model_builder')
+            ->replaceArgument(1, $runtimeServicesLocator);
+    }
+
+    /**
+     * @param class-string $className
+     *
+     * @return list<\ReflectionFunction>
+     */
+    private function formatters(string $className): array
+    {
+        $formatters = [];
+        foreach ((new \ReflectionClass($className))->getProperties() as $property) {
```

Comment:
missing invalid class handling.+ resource tracking

```suggestion
        if (!$r = $container->getReflectionClass($class)) {
            throw new InvalidArgumentException(sprintf('Class "%s" used for service "%s" cannot be found.', $class, $id));
        }
        foreach ($r->getProperties() as $property) {
```